### PR TITLE
[Feature](bangc-ops): add new op deform_roi_pool.

### PR DESCRIPTION
**Pull request by defei-coder**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new operator deform_roi_pool_forward and deform_roi_pool_backward.

## 2. Modification

影响范围/算子：deform_roi_pool_forward、deform_roi_pool_backward
影响版本/分支：master

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

动态阈值 diff1,diff2=[10, 10]

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |        MLU290<br>MLU370<br>MLU590         |
|        2       |Job types                             |          U1       |
|        3       |Layouts                               |          NHWC      |
|        4       |Whether multi-dimensions are supported|              no                        |
|        5       |Whether element zero is supported     |         yes                             |
|        6       |Data type(half/float)                 |           half / float          |
|        7       |Whether there is size limit           |            no                          |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| input、roi、offset、output数据类型一致，并仅支持half和float    |     Normal error    |   [mluOpDeformRoiPoolForward] Check failed: input_desc->dtype == MLUOP_DTYPE_FLOAT || input_desc->dtype == MLUOP_DTYPE_HALF.   Or [mluOpDeformRoiPoolForward] Check failed: input_desc->dtype == output_desc->dtype.                    |
| input、grad_input、grad_output和output仅支持NHWC          |     Normal error    | [mluOpDeformRoiPoolForward] Check failed: input_desc->layout == MLUOP_LAYOUT_NHWC  |
| roi仅支持2维且第二维必须是5 | Normal error | [mluOpDeformRoiPoolForward] Check failed: rois_desc->dims[1] == 5. |
| input和output的最后一维必须相等 | Normal error | [mluOpDeformRoiPoolForward] input channel = 30, output channel = 304, they should be equal. | 
| offset(如果存在)的第两维、output的第二维和pooled_height必须相同 | Normal error | [mluOpDeformRoiPoolForward] Check failed: offset_desc->dims[2] == pooled_height.  | 
| offset(如果存在)的第三维、output的第三维和pooled_width必须相同 | Normal error | [mluOpDeformRoiPoolForward] Check failed: offset_desc->dims[3] == pooled_width.  | 
| roi 的第一维、offset（如果存在）的第一维和 output 的第一维相同 | Normal error | [mluOpDeformRoiPoolForward] Check failed: offset_desc->dims[0] == rois_desc->dims[0].  | 

### 3.2 Accuracy Test
370:
[ SUMMARY  ] Total 582 cases of 2 op(s).
ALL PASSED.
590:
[ SUMMARY  ] Total 582 cases of 2 op(s).
ALL PASSED.
### 3.4 Summary Analysis

All test cases passed.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
                                 const mluOpTensorDescriptor_t output_desc,
                                 void *output);
 
+// Group:DeformRoiPool
+/*!
+ * @brief Computes deformable roi pooling over \b input tensor. This function firstly divides the obtained
+ * candidate region into regions with the same size according to the specified pooling width and pooling height,
+ * then add offsets to rois, and finally calculates the mean value of the sampling points in each bin as output.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6597]: add > adds

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @param[in] offset
+ * Pointer to the MLU memory that stores the offset tensor.
+ * @param[in] pooled_height
+ * An integer value which is the height of the output after pooling.
+ * @param[in] pooled_width
+ * An integer value which is the width of the output after pooling.
+ * @param[in] spatial_scale
+ * A float value which is the scale factor of coordinates of rois.
+ * @param[in] sampling_ratio
+ * An integer value which is the number of sample in one bin. This parameter
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6623]: sample > samples

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @par Formula
+ * - See "DeformRoiPoolForward Operator" section in "Cambricon BANGC OPS User Guide" for details.
+ *
+ * @par Data Type
+ * - The supported data types of \b input, \b rois, \b offset and \b output are as follows:
+ *   Data types of all tensors should be the same.
+ *   - input tensor: half, float.
+ *   - rois tensor: half, float.
+ *   - offset tensor: half, float.
+ *   - output tensor: half, float.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6643]: 这四行去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layouts of \b input, \b rois, \b offset and \b output are as follows:
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - rois tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - offset tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - output tensor: \p MLUOP_LAYOUT_NHWC.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6650]: 这四行去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layouts of grad_output tensor \b grad_output, input tensor \b input, rois tensor \b rois,
+ *   offset tensor \b offset, grad_input tensor \b grad_input and grad_offset tensor \b grad_offset are as follows:
+ *   - grad_output tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - rois tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - offset tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - grad_input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - grad_offset tensor: \p MLUOP_LAYOUT_ARRAY.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6782]: 这六行句尾句号去掉

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - The inputs \b rois and \b offset with NaN or infinity are not supported.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
+ *
+ * @par Reference
+ * - http://github.com/open-mmlab/mmcv/tree/master/mmcv/ops/deform_roi_pool.py
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6816]: https

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ *   - grad_output tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - rois tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - offset tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - grad_input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - grad_offset tensor: \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - The grad_output tensor, input tensor and grad_input tensor must be 4D.
+ * - the sizes of the lowest dimension of grad_output tensor, input tensor and grad_input tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6786]: The

Comment (defei-coder) [reply bangc-ops/mlu_op.h:6786]: Modified

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6816]: Modified

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6782]: Modified

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * @par Data Type
+ * - The supported data types of grad_output tensor \b grad_output, input tensor \b input, rois tensor \b rois,
+ *   offset tensor \b offset, grad_input tensor \b grad_input and grad_offset tensor \b grad_offset are as follows:
+ *   Data types of all tensors should be the same.
+ *   - grad_output tensor: half, float.
+ *   - input tensor: half, float.
+ *   - rois tensor: half, float.
+ *   - offset tensor: half, float.
+ *   - grad_input tensor: half, float.
+ *   - grad_offset tensor: half, float.
```
[the remark this answers is not in the document]
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6772]: Modified

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * An integer value which is the number of sample in one bin. This parameter
+ * only works when it is greater than zero.
+ * @param[in] gamma
+ * A float value which is the scale factor of offset.
+ * @param[in] grad_input_desc
+ * The descriptor of grad_input tensor, which contains the dimension and layout of grad_input tensor.
+ * @param[out] grad_input
+ * Pointer to the MLU memory that stores the gradient of the input tensor.
+ * @param[in] grad_offset_desc
+ * DThe descriptor of grad_offset tensor, which contains the dimension and layout of grad_offset tensor.
```
[the remark this answers is not in the document]
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6754]: Modified

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+            pooled_height: 1.0, pooled_width: 1.0, spatial_scale: 1.0,
+            sampling_ratio: 1, gamma: 1
+
+     output array by 1  1  1 * 1 -->
+         output: [[[[2.25]]]]
+     @endverbatim
+ *
+ *
+ * @par Reference
+ * - http://github.com/open-mmlab/mmcv/tree/master/mmcv/ops/deform_roi_pool.py
```
[the remark this answers is not in the document]
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6695]: Modified

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6650]: Modified

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6643]: Modified

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * A float value which is the scale factor of offset.
+ * @param[in] output_desc
+ * The descriptor of output tensor, which contains the dimension and layout of output tensor.
+ * @param[out] output
+ * Pointer to the MLU memory that stores the output tensor.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_NOT_SUPPORTED
+ *
+ * @par Formula
+ * - See "DeformRoiPoolForward Operator" section in "Cambricon BANGC OPS User Guide" for details.
```
[the remark this answers is not in the document]
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6635]: Modified

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6623]: Modified

File: bangc-ops/mlu_op.h (hunk 1 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:6597]: Modified

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -6590,6 +6590,250 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+ * - The shape of \b rois should be [rois_num, 5].
+ * - The shape of \b offset should be [rois_num, 2, pooled_height, pooled_width].
+ * - The shape of \b output should be [rois_num, pooled_height, pooled_width, channels].
+ * - \b rois[i] consists of [batch_id, x1, y1, x2, y2]. \p batch_id should be in the range of [0, batch_num - 1].
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @par Performance Optimization
+ * - None.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6669]: 删除这个字段

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h]: Modified

**Review by GraceGuanH — APPROVED** [review]

File: docs/bangc-docs/design_docs/deform_roi_pool_backward/deform_roi_pool_backward.md [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+
+1、bangc 代码中加入必要的 log 信息，比如输入的规模、数据类型、layout 这些，以及如果出错会导致程序 core dump 的变量，比如 IO 指令的 data_size、dim xyz 的值等，这些信息都是有利于快速定位问题；
+
+2、对每一个函数命名变量命名都有充分的注释；,
+
+3、避免魔鬼数字，对于确定的数字尽量使用公共宏来替代。
+
+### 3.7 测试用例设计
+
+- 框架在需求列表中给出的算子在网络中用到的规模：
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/deform_roi_pool_backward/deform_roi_pool_backward.md:330]: 该算子在网络中用到的规模：

**Review by mahxn0 — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 14]
```diff
@@ -355,3 +355,33 @@ mluOpActiveRotatedFilterForward
 ----------------------------------
 该算子根据位置信息对输入进行旋转。这个算子编码方向信息并生成方向敏感特征。
 
+.. _deform_roi_pool_backward:
```
Comment (WangYuan-cn) [inline docs/bangc-docs/user_guide/9_operators/index.rst:358]: 这句是release note 超连接，建议删除

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -7191,6 +7191,248 @@ mluOpActiveRotatedFilterForward(const mluOpHandle_t handle,
...
+
+     param:
+            pooled_height: 1.0, pooled_width: 1.0, spatial_scale: 1.0,
+            sampling_ratio: 1, gamma: 1
+
+     output array by 1  1  1 * 1 -->
+         output: [[[[2.25]]]]
+     @endverbatim
+ *
+ *
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:7292]: 多了行空

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**